common/kvstore/src/test/java/org/apache/spark/util/kvstore/LevelDBSuite.java

```diff
@@ -383,6 +385,51 @@ public void testSkipAfterDBClose() throws Exception {
     assertFalse(iter.skip(1));
   }
 
+  @Test
+  public void testResourceCleaner() throws Exception {
+    File dbPathForCleanerTest = File.createTempFile(
+      "test_db_cleaner.", ".rdb");
+    dbPathForCleanerTest.delete();
+
+    LevelDB dbForCleanerTest = new LevelDB(dbPathForCleanerTest);
+    try {
+      for (int i = 0; i < 8192; i++) {
+        dbForCleanerTest.write(createCustomType1(i));
+      }
+      LevelDBIterator<CustomType1> levelDBIterator = (LevelDBIterator<CustomType1>) dbForCleanerTest.view(CustomType1.class).iterator();
+      Reference<LevelDBIterator<?>> reference = getRocksDBIteratorRef(levelDBIterator, dbForCleanerTest);
+      assertNotNull(reference);
+      DBIterator it = levelDBIterator.internalIterator();
+      // it has not been closed yet, hasNext execute normally.
+      assertTrue(it.hasNext());
+      // Manually set rocksDBIterator to null, to be GC.
+      levelDBIterator = null;
+      // 100 times gc, the rocksDBIterator should be GCed.
+      int count = 0;
+      while (count < 100 && !reference.refersTo(null)) {
+        System.gc();
+        count++;
+        Thread.sleep(100);
+      }
+      // check rocksDBIterator should be GCed
+      assertTrue(reference.refersTo(null));
+      // Verify that the Cleaner will be executed after a period of time, there is an exception message. java.lang.AssertionError: This object has been deleted
+      assertThrows(AssertionError.class, it::hasNext, "This object has been deleted");
```

Comment:
Can we change to `JniDBIterator` to get the corresponding `NativeIterator` object and check the result of `NativeIterator.isAllocated`? This might require getting `NativeIterator` through reflection?

Reply:
Done.